apps/escrow/src/hooks/program/useUploadProgram.ts

```diff
@@ -0,0 +1,195 @@
+import { web3FromSource } from '@polkadot/extension-dapp';
+import { useApi, useAccount, useAlert, DEFAULT_ERROR_OPTIONS, DEFAULT_SUCCESS_OPTIONS } from '@gear-js/react-hooks';
+import { GearApi, Hex, MessageEnqueued, MessagesDispatched, Metadata, ProgramChanged } from '@gear-js/api';
+import { UnsubscribePromise } from '@polkadot/api/types';
+import { Event } from '@polkadot/types/interfaces';
+import { useEscrowMeta, useEscrowOpt } from '../api';
+
+enum ProgramStatus {
+  Success = 'success',
+  Failed = 'failed',
+  InProgress = 'in progress',
+}
+
+export interface UploadProgramModel {
+  id?: string;
+  meta?: Metadata;
+  value: number;
+  title?: string;
+  gasLimit: number;
+  initPayload: string;
+  programName?: string;
+  payloadType?: string;
+}
+
+enum TransactionStatus {
+  Ready = 'Ready',
+  InBlock = 'InBlock',
+  IsInvalid = 'IsInvalid',
+  Finalized = 'Finalized',
+}
+
+const PROGRAM_ERRORS = {
+  UNAUTHORIZED: 'Unauthorized',
+  INVALID_PARAMS: 'Invalid method parameters',
+  INVALID_TRANSACTION: 'Transaction error. Status: isInvalid',
+  PROGRAM_INIT_FAILED: 'Program initialization failed',
+  GEAR_NODE_ERROR: 'Gear node error',
+  BALANCE_LOW: 'Invalid transaction. Account balance too low',
+  PAYLOAD_ERROR: 'payload.toHex is not a function',
+};
+
+const getExtrinsicFailedMessage = (api: GearApi, event: Event) => {
+  const { docs, method } = api.getExtrinsicFailedError(event);
+  const formattedDocs = docs.filter(Boolean).join('. ');
+
+  return `${method}: ${formattedDocs}`;
+};
+
+const waitForProgramInit = (api: GearApi, programId: string) => {
+  let messageId: Hex;
+  let unsubPromise: UnsubscribePromise;
+
+  const unsubscribe = async () => (await unsubPromise)();
+
+  return new Promise<string>((resolve) => {
+    unsubPromise = api.query.system.events((events) => {
+      events.forEach(({ event }) => {
+        switch (event.method) {
+          case 'MessageEnqueued': {
+            const meEvent = event as MessageEnqueued;
+
+            if (meEvent.data.destination.eq(programId) && meEvent.data.entry.isInit) {
+              messageId = meEvent.data.id.toHex();
+            }
+
+            break;
+          }
+
+          case 'MessagesDispatched': {
+            const mdEvent = event as MessagesDispatched;
+
+            mdEvent.data.statuses.forEach(
+              (status, id) => id.eq(messageId) && status.isFailed && resolve(ProgramStatus.Failed),
+            );
+
+            break;
+          }
+
+          case 'ProgramChanged': {
+            const pcEvent = event as ProgramChanged;
+
+            if (pcEvent.data.id.eq(programId) && pcEvent.data.change.isActive) {
+              resolve(ProgramStatus.Success);
+            }
+
+            break;
```

Comment:
```suggestion
          case 'MessagesDispatched': {
            const mdEvent = event as MessagesDispatched;

            mdEvent.data.statuses.forEach(
              (status, id) => {
                if (id.eq(messageId)) { 
                  status.isFailed && resolve(ProgramStatus.Failed);
                  status.isSuccess && resolve(ProgramStatus.Success);
                }
              }
            );
            break;
          }
```
You can do it as above